First draft of 0026-activation-distribution.md (#26)
openmrs/openmrs-rfc-frontend: text/0026-activation-distribution.md

1. Have a proper root module in the microfrontend modules, which defines the
activity function and lazy loads the rest of the microfrontend (via bundle
splitting).
2. Export special functionality that correctly identifies the module as a
microfrontend that can be properly integrated.

The process to migrate from a centralized module containing all activity
functions to a distributed starts with a new `index.ts` in the microfrontends.

The `index.ts` may be as simple as follows:

```js
// set the public path to be able to retrieve side bundles
import "./set-public-path";
// register the application in the shared single-spa package
import { registerApplication } from "single-spa";
// get useful helpers for the activity functions from the loaded root-config
import { routePrefix } from '@openmrs/esm-root-config';

// perform the registration with the lazy loading of the lifecycle and its activity function
registerApplication(
  "@openmrs/esm-home",
  () => import('./openmrs-esm-home'),
  location => routePrefix("home", location),
);

// export other dependencies to be used, e.g., by the dev tooling
export { backendDependencies } from "./openmrs-backend-dependencies";

> ```

Reviewer: Also, I just want to make sure that we're on the same page that this no longer applies; that index.ts should satisfy the interface that you call setupOpenMRS and I call the "microfrontend interface," and it should not call registerApplication. Yeah?

Author: Good catch! I'll correct it!

# Project title
- Start Date: 2020/04/27
- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/26

## Decision, including impact on distributions
OpenMRS Microfrontends will have their own default activator functions.
Microfrontends will be identified as such by their names being suffixed with `-app`.
Microfrontends' will provide activator and lifecycle functions in their exports according
to a specific microfrontend interface, defined in  this RFC. The interface bundle will be
dynamically split from the main application bundle via a dynamic import in the lifecycle function.
The app shell will load the app shell config if available. The app shell will register all
microfrontends in the import map with Single SPA, and start the application.

> The existing concept of a "root config" will be superseded by the app shell
> and an app shell config.

Reviewer: ```suggestion
The existing concept of a "root config" will be superseded by the initial script,
which is optionally configurable by a new kind of "root config."
```

Reviewer: @FlorianRappl Shoot, just this one more, and then that's it!

Author: Awesome thanks for the enhancement @brandones ! :rocket:

# Project title
- Start Date: 2020/04/27
- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/26

## Decision, including impact on distributions
OpenMRS Microfrontends will have their own default activator functions.

> Microfrontends will be identified as such by their exports satisfying a
> specific microfrontend interface, defined in this RFC. The app shell will load
> the app shell config if available. The app shell will register all
> microfrontends in the import map with Single SPA, and start the application.

Reviewer: ```suggestion
Microfrontends will be identified as such by their names being suffixed with `-app`.
Microfrontends' will provide activator and lifecycle functions in their exports according
to a specific microfrontend interface, defined in  this RFC. The interface bundle will be
dynamically split from the main application bundle via a dynamic import in the lifecycle function.
The app shell will load the app shell config if available. The app shell will register all
microfrontends in the import map with Single SPA, and start the application.
```

The existing concept of a "root config" will be superseded by the app shell
and an app shell config.

In order to implement this change, we will need to

1. Add index.ts entrypoint files to each microfrontend that satisfy the
microfrontend interface (exporting lifecycle and activate).
2. Change the initial script to behave as described.
3. Migrate all existing root configs to the new structure. This will generally
just mean removing the calls to registerApplication and exporting those
activators that are different from the defaults.

> Instead of lazy loading all modules from the importmaps the modules are
> directly evaluated. 
> 

Reviewer: ```suggestion
```

## Definition
An **activation function** is a JavaScript function that is used by the Single
SPA library to determine if and when a microfrontend should be activated.

The **activation of a microfrontend** describes when the lifecycle of an app
should begin.

The **lifecycle** of an app determines what to do technically for booting,
mounting, updating, and unmounting an application in the DOM of the currently
presented website.

The **DOM** refers to the API responsible for the object representation of a
website in the browser.

> An **app shell** refers to the index.html page that includes the required base
> scripts. Thus triggering the whole loading process necessary for all the
> microfrontends.

Reviewer: ```suggestion
An **app shell** refers to the index.html page that includes required libraries
and the *initial script*.
```

> The **initial script** is the script performing the loading of the Single SPA
> library. It sets up important globals and imports the important global modules.

Reviewer: ```suggestion
The **initial script** is the script which identifies microfrontends in the import map,
loads them, loads the *root config*, and registers microfrontends with the Single SPA
application. It sets up important globals and imports the important global modules.
```

The new **root config** replaces the old concept of a [root config][1]. An

> optional element of the import map, named `@openmrs/esm-root-config`, which can

Reviewer: ```suggestion
optional element of the import map, named `root-config`, which can
```

My understanding is that `@openmrs/esm-root-config` will either cease to exist or be converted into the *initial script*, right? This new "root config" (maybe we should call it something else?) will be totally different. There will be no default version of it, it's totally optional and implementation-specific. So it wouldn't make sense to namespace it to `@openmrs`. Does that make sense?

be used to (1) provide module configs to `@openmrs/esm-module-config`, (2)
override module activator functions, and (3) do any other pre-app-launch setup

> wanted by the actual implemention. Removed responsibility: Registration of
> microfrontends with the Single SPA framework.

Reviewer: ```suggestion
wanted by the actual implemention. It does not register
microfrontends with the Single SPA framework.
```

> A **microfrontend interface** describes the conditions on the exports of an ESM
> which, if met, indicate that the module is an OpenMRS microfrontend module. In
> particular, the modules exporting a `setupOpenMRS` are considered a
> microfrontend module.
> 
> The interface for the `setupOpenMRS` function can be described in TypeScript as

Reviewer: ```suggestion
The **microfrontend interface** is the interface that the *initial script* expects microfrontends
to satisfy. Microfrontends are ESMs whose names are suffixed with `-app`. The interface
consists of a single function, `setupOpenMRS`:
```

```ts
interface SetupOpenMRS {
  (): {

> 	  lifecycle: LifeCycles<T> | (() => Promise<LifeCycles<T> | Splat<LifeCycles<T>>>);
>     activate(location: Location): boolean;

Reviewer: ```suggestion
    lifecycle: LifeCycles<T> | (() => Promise<LifeCycles<T> | Splat<LifeCycles<T>>>);
    activate(location: Location): boolean;
```

  };
}
```

The evaluation of the exports of the modules from the importmaps can be done in
the initial script of the app shell.

The process to migrate from a centralized module containing all activity
functions to a distributed introduces a new `index.ts` in the microfrontends.

The `index.ts` may be as simple as follows:

```js
// set the public path to be able to retrieve side bundles
import "./set-public-path";
// get useful helpers for the activity functions from the loaded root-config
import { routePrefix } from '@openmrs/esm-root-config';

export function setupOpenMRS() {
  return {

>     lifecycle: () => import('./openmrs-esm-home'),

Author: @joeldenning -> side-bundle.

    activate: location => routePrefix("home", location),
  };
}

// export other dependencies to be used, e.g., by the dev tooling
export { backendDependencies } from "./openmrs-backend-dependencies";
```

## Reason for decision
- Enhance possibility of distributing the microfrontends
- Make it possible to create self-containing microfrontends
- Allow an easy packaging model for OpenMRS distributions
- Use existing standards instead of creating new ones
- Cooperate with existing architecture
- Improve developer experience

## Alternatives
There are three main alternatives that can be seen:

1. Switching away from Single SPA to microfrontend frameworks that support the
notion of distributed microfrontends.
2. Keeping the original notion of a single file that has the ultimate knowledge
(with all the drawbacks and complexity for distribution).

> 3. Using a service to get the information via an API call. Besides the original

Reviewer 2: As of single-spa@5.3.0, single-spa accepts strings as activity functions. For example:

```js
singleSpa.registerApplication({
  name: 'nav',
  app: () => System.import('nav'),
  activeWhen: ['/']
})

singleSpa.registerApplication({
  name: 'patient-chart',
  app: () => System.import('patient-chart'),
  activeWhen: ['/patient/:uuid/chart']
})
```

- [Release notes](https://github.com/single-spa/single-spa/releases/tag/v5.3.0)
- [Documentation](https://single-spa.js.org/docs/configuration#configactivewhen)

^ I think that storing a root config in a service/database is a very viable approach, and is one that I know several organizations are already employing. The database could store the name of the service and a list of activeWhen strings - that is all that is needed to register the application.

Author: > If I understand this proposal correctly, I think that it would require downloading the code for all microfrontends in order to register the single-spa applications?

No - the microfrontends are bundle splitted as stated; so you only download essentially 0.5 KB - not the whole MF. So I assume this was misunderstood.

> As of single-spa@5.3.0, single-spa accepts strings as activity functions.

That's great and certainly useful also for the JSON configuration. However, it does not help with the distribution.

> I think I lean toward Alternative 3

Wonderful - thanks for the suggestion! :)

Reviewer 2: > No - the microfrontends are bundle splitted as stated; so you only download essentially 0.5 KB - not the whole MF.

I've read through the proposal several times and do not see where it discusses code splitting. Could you help me understand this? Even with code splitting, this still comes with a performance cost - if there are 25 microfrontends, this proposal requires downloading the entry bundle for all 25 of them in order to register the applications?

Reviewer 2: > That's great and certainly useful also for the JSON configuration. However, it does not help with the distribution.

Could you help me understand this? Distributions already have a variety of configuration stored in the database that is accessible via the Admin tool. I believe that this includes which Java modules are activated or not, which is fairly similar to which javascript modules are activated. I agree that storing code in a database has downsides, but don't see how it prevents distribution, since my understanding of creating OpenMRS distributions is that this often involves setting up a database with correct config.

Author: > if there are 25 microfrontends, this proposal requires downloading the entry bundle for all 25 of them in order to register the applications?

Right now we have multiple loose scripts via importmaps, which could also be served as a single UMD. You cannot argue in favor of fixed importmaps (+ include 6 loose SystemJS scripts, all which could be served as a single UMD, too), but not be in favor of microfrontends being loosely coupled. Microfrontends are all about loose coupling.

Furthermore, this only needs to happen once. The script can then be cached.

I am not sure if I read all that correctly - to me it reads with a certain attitude (maybe I am mistaking here). I propose to discuss this in a call. I guess it would make more sense to you then.

Reviewer 2: My intent here is engaging in a respectful, productive dialog. I am not arguing with you, but seeking to understand. Thanks for the [clarification comment](https://github.com/openmrs/openmrs-rfc-frontend/pull/26#discussion_r420865829) below calling out the code split via `import()`, that helps me understand how this proposal works 👍.

> You cannot argue in favor of fixed importmaps (+ include 6 loose SystemJS scripts, all which could be served as a single UMD, too), but not be in favor of microfrontends being loosely coupled. Microfrontends are all about loose coupling.

I don't view putting top level routes into a root config as tight coupling. And do not see the connection between loading SystemJS scripts and the current discussion here.

Moving the activity function logic into the microfrontend itself does make it more self encapsulated. But I think that we shouldn't sacrifice performance for that encapsulation.

#### Summary of my concern

The current proposal requires that every microfrontend's entry bundle be downloaded before the first meaningful render. The entry bundles can be quite small due to code splits, but doing this is still is a fairly large and dynamic number of network requests before the first meaningful render. I think that a fairly large, dynamic number of network requests has a performance penalty that should be avoided if possible.

#### Maybe another alternative?

Perhaps instead of downloading all modules in the browser in order to retrieve their activity functions, each package could publish to npm a separate file that could be included into the root-config as a build-time dependency?

Example:

```js
// root-config.js

// these are "build-time" modules that come from node_modules
import registerPatientChart from '@openmrs/esm-patient-chart/registration'
import registerLogin from '@openmrs/esm-login/registration'

registerPatientChart()
registerLogin()
```

```js
// patient-chart/dist/registration.js
import { registerApplication } from 'single-spa';

export default function () {
  registerApplication({
    name: '@openmrs/esm-patient-chart',
    app: () => System.import('@openmrs/esm-patient-chart'),
    activeWhen: ['/patient/:uuid/chart']
  })
}
```

Author: Let me try to show you the connection to SystemJS (which is just one example). You refer about saving requests, which obviously contradict the current design (e.g., instead of a single bundle we serve 7 scripts, 6 references, 1 inline). I would love to set the same metrics everywhere.

> each package could publish to npm a separate file that could be included into the root-config as a build-time dependency?

Here I think there is a misunderstanding. You have several independently developed and deployed modules. How can they now come together in a build (coming from a system that is shipped pre-built and does not have any knowledge of which and how many modules are actually used later on)?

The idea of this proposal is to allow a free composition.

Reviewer 2: > You refer about saving requests, which obviously contradict the current design (e.g., instead of a single bundle we serve 7 scripts, 6 references, 1 inline). I would love to set the same metrics everywhere.

The difference between the systemjs scripts and this proposal is that the systemjs scripts are a fixed 5-6 requests, whereas the number of applications is dynamic, and could be as high as 25 or even 50 (for a large organization). The systemjs library itself is quite small (~5kb gzipped for everything).

A dynamic number of requests has a lot more potential to impact performance than a fixed number of requests.

> Here I think there is a misunderstanding. You have several independently developed and deployed modules. How can they now come together in a build (coming from a system that is shipped pre-built and does not have any knowledge of which and how many modules are actually used later on)?

Most OpenMRS distributions do not independently deploy modules, but rather package them all up and ship them out as `.war` files. In order for deployments to be possible, there must be centralization of servers, which is not true of the majority of the OpenMRS community. As far as I know, there are only very few organizations (Ampath being one of them) that have centralized their servers.

I agree that my proposal disallows for independent deployment, which is not ideal. That's a valid criticism and perhaps the idea should be scrapped 😄.

My intent here is to try to find a way to avoid an unknown and dynamic number of network requests before the first meaningful render.

Author: > My intent here is to try to find a way to avoid an unknown and dynamic number of network requests before the first meaningful render.

I think this is good and obviously the argument with static vs dynamic number is also true. Nevertheless, the argument that these initial requests should be already bundled is in the same sense also right - and justifies that we are not playing with the same rules here. Unfortunately.

Nevertheless, coming back to the topic. The thinking with the unknown number of requests does not take into consideration HTTP/2 batching and multiplexing. The cost of multiple requests for scripts that are ~less than 1 KB in size is in reality quite negligible. And again - that these requests exist is only true for FTU. Afterwards, they are all there - which by the way also improves update scenarios. Instead of two things you'll only need to update one module. In this case only the updated single script is requested from the browser.

Reviewer 2: Fair enough - if it is ensured that the code splitting occurs, http2 enabled, gzip, etc, perhaps the performance impact is worth the benefits of moving the top level routes to the microfrontends themselves.

I'll leave this conversation here for now and let the MF core team evaluate this PR further. I am not as involved in openmrs as I was last year and view my role here as just providing discussion points / insight, but not as the decision maker.

I think you're a smart person @FlorianRappl who clearly has some experience and knows your stuff. I'm glad OpenMRS has you to help push things forward.
